scripts/migrations/2020-10-miro-migration/transfer_packager.py

```diff
@@ -0,0 +1,195 @@
+    file_location = os.path.join(target_folder, file_name)
+
+    s3_content_length = get_s3_content_length(
+        s3_client=s3_client, s3_bucket=s3_bucket, s3_key=s3_key
+    )
+
+    def _download():
+        required_dir = os.path.dirname(file_location)
+        os.makedirs(required_dir, exist_ok=True)
+        s3_client.download_file(Bucket=s3_bucket, Key=s3_key, Filename=file_location)
+
+    # Check if we already have this file
+    if os.path.isfile(file_location):
+        # If the length doesn't match overwrite
+        if not os.path.getsize(file_location) == s3_content_length:
+            _download()
+    else:
+        _download()
+
+    _check_local_content_length(
+        file_location=file_location, expected_content_length=s3_content_length
+    )
+
+
+def _download_objects_from_s3(s3_client, target_folder, s3_bucket, s3_key_list, prefix):
+    def _get_s3_object(s3_key):
+        _download_s3_object(
+            s3_client=s3_client,
+            s3_bucket=s3_bucket,
+            s3_key=s3_key,
+            target_folder=target_folder,
+            prefix=prefix,
+        )
+
+    # itertools below expects an iterator
+    s3_key_list_iter = iter(s3_key_list)
+
+    with tqdm(total=len(s3_key_list)) as pbar:
+        # From https://alexwlchan.net/2019/10/adventures-with-concurrent-futures/
+        with concurrent.futures.ThreadPoolExecutor() as executor:
+            # Schedule the first N futures.  We don't want to schedule them all
+            # at once, to avoid consuming excessive amounts of memory.
+            futures = {
+                executor.submit(_get_s3_object, task): task
+                for task in itertools.islice(s3_key_list_iter, S3_DOWNLOAD_CONCURRENCY)
+            }
+
+            while futures:
+                # Wait for the next future to complete.
+                done, _ = concurrent.futures.wait(
+                    futures, return_when=concurrent.futures.FIRST_COMPLETED
+                )
+
+                for fut in done:
+                    futures.pop(fut)
+
+                pbar.update(len(done))
+
+                # Schedule the next set of futures.  We don't want more than N futures
+                # in the pool at a time, to keep memory consumption down.
+                for task in itertools.islice(s3_key_list_iter, len(done)):
+                    fut = executor.submit(_get_s3_object, task)
+                    futures[fut] = task
+
+        target_folder_count = sum(
+            [len(files) for r, d, files in os.walk(target_folder)]
+        )
+
+        assert target_folder_count == len(s3_key_list), (
+            f"Unexpected file count in {target_folder}: "
+            f"{target_folder_count} != {len(s3_key_list)}"
+        )
+
+
+def _compress_folder(target_folder, remove_folder=True):
+    archive_name = shutil.make_archive(target_folder, "zip", target_folder)
+    if remove_folder:
+        shutil.rmtree(target_folder, ignore_errors=True)
```

Comment:
minor: You only call `_compress_folder` once, and you don't override `remove_folder`. I would be explicit, remove the parameter, don't hide this behind a conditional.

Reply:
It's there as a convenience for me to see the folder if I need to ... but I guess it's as easy to just comment this line out if I need to.